pkg/prometheus/agent/operator.go

```diff
@@ -605,10 +605,35 @@ func (c *Operator) sync(ctx context.Context, key string) error {
 		return fmt.Errorf("synchronizing web config secret failed: %w", err)
 	}
 
-	// Create governing service if it doesn't exist.
+	// If a ServiceName is specified, check that the service exists in the namespace. If it does not exist, fail
+	// the reconciliation.
+	// If the ServiceName is not specified, create a governing service if it doesn't exist.
+	// Also, ensure that the Prometheus instance is selected by the service. If not, fail the reconciliation.
+	// TODO[mviswanathsai]: is it possible to refactor this (and other) code repetitions between the Prometheus server and agent?
```

Comment:
yes please :)
side suggestion: please do in a preliminary PR.

Reply:
Should we complete this PR and then move onto the refactor? Seems like the refactor might be a bit of work.

Reply:
I'd rather do it before this PR otherwise there's a risk we never get to it.

Reply:
Got it!

Reply:
Can you give me pointers on where to start with the refactor? @simonpasquier

Reply:
Sorry I was confused: it should be done in this PR. The logic when `p.Spec.ServiceName != nil` should be similar between agent and server.